BG96: Add segmentation to TCP socket send

boraozgen opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Unlike some other cellular modem drivers which use segmentation, the BG96 driver did not support sending of large TCP packets, i.e. maximum TCP packet size was restricted to 1400 bytes. This PR adds segmentation to the send function, so larger packets can be sent using the TCP socket.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [X] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

pan- commented on connectivity/drivers/cellular/QUECTEL/BG96/QUECTEL_BG96_CellularStack.cpp:
```diff
@@ -358,29 +354,57 @@ nsapi_size_or_error_t QUECTEL_BG96_CellularStack::socket_sendto_impl(CellularSoc
-    _at.resp_stop();
-    if (strcmp(response, "SEND OK") != 0) {
-        return NSAPI_ERROR_DEVICE_ERROR;
+        while ((count > 0)) {
```
It would be worth handling the zero length packet case.

boraozgen replied:
This wasn't handled before either. What is the expected behaviour in this case?

It should return 0 in this case as it will skip the AT write part completely. We could do an argument validation at the beginning of the function, but I believe it is not necessary.

pan- replied:
Zero length packets are valid and should be _written_. That was the case before, the function was not exiting if the size in input was 0. With the change, nothing is written because the loop is not even entered if `size = 0`.
I suggest the use of a `do { } while(...);` construct instead of a `while`.

boraozgen replied:
Didn't know that, thanks. Done.

kjbracey replied:
Going to contradict @pan- here. Zero-length packets are a meaningless concept for TCP, because there are no packets. It's a stream. Writing 0 bytes is a straight no-op in a lot of implementations. Sometimes it might give you back a pending error indication, including a potential "connection reset" but a zero-length read is a more common idiom to get that, because a fair few write implementations exit early on len=0.

Zero-length packets are valid and meaningful for UDP, and those should be passed to and from the underlying modem (in the hope that it will support them - many don't).

But it would be acceptable for this layer to silently bin 0-length TCP writes. Indeed, it would be the best thing to do if you knew that the modem went weird and returned an error rather than "0". But if this modem seems to work fine and return 0 for a 0-length write in the normal case, you may as well pass it through, as it now does,

boraozgen replied:
I did not test if the modem handles this case.

Regarding zero-length packets, I have read that they might be used to keep the TCP connection alive. Ref: https://osqa-ask.wireshark.org/questions/11863/why-do-tcp-clients-send-packets-with-no-data/

pan- replied:
Thanks @kjbracey-arm for providing more insight. I agree it doesn't makes much sense to send a zero length tcp packet if the ACK flag isn't controlled. 

I went through the application note of the module; `AT+QISEND=<connectID>,0` is supported for the _keepalive_ use case but it won't work with the current implementation as the expected return code is `OK` and not `SEND OK`. It also returns the length of sent data, acknowledged data and unacknowledged data. 

I think the best option for such use case would be to support `NSAPI_KEEPALIVE` using the `AT+QICFG="tcp/keepalive"` AT command in `setsockopt` of the module. 

@boraozgen My apologies for the confusion, I think it is best to revert back to the `while` loop instead of `do { } while()` and drop 0 length packet until a better implementation is made.

kjbracey replied:
> Regarding zero-length packets, I have read that they might be used to keep the TCP connection alive.

Certainly there can be zero-data segments under the hood in TCP, for a variety of reasons, but from the API point of view, it's a stream. You can't do anything to (reliably) influence segment size via the read/write API.

As pan- says, keep-alives have to be configured via a separate control.

kjbracey replied:
Interesting to read that that modem gives "send 0" a special function. That does make sense - writing 0 in the usual sense is meaningless, so may as well assign a meaning to the command.

For the BSD/POSIX `write`/`send` API though, the "be a no-op" for 0 is required by the general "stream" semantics. It has to be a no-op for all stream-type devices - UART/TCP/files...

boraozgen replied:
> @boraozgen My apologies for the confusion, I think it is best to revert back to the `while` loop instead of `do { } while()` and drop 0 length packet until a better implementation is made.

Reverted.

pan- reviewed (changes requested):
Thanks for your submission @boraozgen . However the case where the length of the TCP packet payload is 0 is not handled by the modifications. 
Can you add support for it ?

boraozgen committed:
BG96: Add segmentation to TCP socket send

Unlike some other cellular modem drivers which use
segmentation, the BG96 driver did not support sending of large
TCP packets, i.e. maximum TCP packet size was restricted to
1400 bytes. This adds segmentation to the send function, so
larger packets can be sent using the TCP socket.
